### ISSUE TITLE
False Positive with Style/CollectionCompact on Hash

### ISSUE THREAD

**Issue by disk42** [rubocop/rubocop#9611]
There are times when using `compact` on a Hash is not appropriate, and `reject` is, as they can return different results:

```
{ key1: :foo, key2: false, key3: nil }.compact == { key1: :foo }

{ key1: :foo, key2: false, key3: nil }.reject { |_, value| value.nil? } == { key1: :foo, key2: false }
```

There are times where the second one is the desired output (and `reject` must be used), but Rubocop flags it.

--------

## Expected behavior

To not flag `reject` on a hash in the situation described above.

## Actual behavior

Rubocop indicates that `compact` should be used, even though it returns a different result than `reject`.

## RuboCop version

```
$ bundle exec rubocop -V
1.9.1 (using Parser 3.0.0.0, rubocop-ast 1.4.1, running on ruby 2.7.2 x86_64-linux)
  - rubocop-performance 1.9.2
  - rubocop-rails 2.9.1
```

**Comment by dvandersluis** [comment]
This doesn't seem to be valid:

```ruby
{ key1: :foo, key2: false, key3: nil }.compact
#=> {:key1=>:foo, :key2=>false}
```

I've tested in ruby 2.6, 2.7 and 3.0.
